File: src/main/java/com/mycompany/myapp/service/LaunchDarklyService.java
package com.mycompany.myapp.service;

import com.launchdarkly.sdk.LDContext;
import com.launchdarkly.sdk.server.LDClient;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

/**
 * 🎯 DEMO: Backend LaunchDarkly Integration
 * 
 * This service demonstrates FLAG #2: quick-test-voice-chat-enabled
 * 
 * What it does:
 * - Receives conversation ID from frontend via X-Conversation-ID header
 * - Creates a LaunchDarkly context of kind "conversation" (NOT "user")
 * - Evaluates the 'quick-test-voice-chat-enabled' flag based on conversation ID pattern
 * - Returns boolean: true (voice enabled) or false (voice disabled)
 * 
 * KEY PRIVACY FEATURE:
 * - Backend NEVER receives user email or personal data
 * - Only receives conversation ID like "premium_conv_abc12345"
 * - Flag targeting rules check the prefix pattern (e.g., "starts with premium_")
 * 
 * This demonstrates frontend/backend flag coordination with complete data separation!
 */
@Service
public class LaunchDarklyService {

    private static final Logger log = LoggerFactory.getLogger(LaunchDarklyService.class);
    
    // 🎯 FLAG #2: Server-side voice chat feature flag
    private static final String VOICE_CHAT_FLAG = "quick-test-voice-chat-enabled";

    private final LDClient ldClient;

    public LaunchDarklyService(LDClient ldClient) {
        this.ldClient = ldClient;
    }

    /**
     * 🎯 DEMO STEP 3: Backend Flag Evaluation
     * 
     * Evaluates the voice-chat-enabled flag using the conversation ID as context.
     * 
     * @param conversationID The conversation ID from X-Conversation-ID header
     *                       Examples: "premium_conv_abc123", "test_conv_xyz789", "conv_def456"
     * @return true if voice is enabled for this conversation tier, false otherwise
     */
    public boolean evaluateVoiceChat(String conversationID) {
        if (ldClient == null) {
            log.warn("⚠️ LaunchDarkly client not available. Returning false for voice-chat-enabled flag.");
            log.warn("⚠️ Configure LD_SDK_KEY environment variable to enable flag evaluation.");
            return false;
        }

        try {
            // Create LaunchDarkly context with "conversation" kind
            // This is DIFFERENT from the frontend "user" context!
            LDContext context = LDContext.builder(conversationID)
                .kind("conversation")  // Important: kind = "conversation" not "user"
                .build();

            // 🎯 FLAG #2: Evaluate 'quick-test-voice-chat-enabled'
            boolean flagValue = ldClient.boolVariation(VOICE_CHAT_FLAG, context, false);

            log.info("✅ Flag evaluated: {} for conversationID '{}' = {}", 
                VOICE_CHAT_FLAG, conversationID, flagValue);
            
            log.debug("📋 Context details - Key: {}, Kind: {}", 
                context.getKey(), context.getKind());

            return flagValue;
        } catch (Exception e) {
            log.error("❌ Error evaluating LaunchDarkly flag for conversationID: {}", conversationID, e);
            return false;
        }
    }
}

File: src/main/webapp/app/services/launchdarkly.service.ts
import { Injectable } from '@angular/core';
import * as LDClient from 'launchdarkly-js-client-sdk';
import { nanoid } from 'nanoid';
import { BehaviorSubject, Observable } from 'rxjs';
import { environment } from 'environments/environment';

/**
 * 🎯 DEMO: Frontend LaunchDarkly Integration
 * 
 * This service demonstrates FLAG #1: conversationid-prefix
 * 
 * What it does:
 * - Evaluates the 'conversationid-prefix' flag based on user attributes (email, beta)
 * - Returns a tier prefix: "premium", "test", "beta", or "" (empty string)
 * - Generates a conversation ID like: "premium_conv_abc12345"
 * - This ID is sent to the backend via HTTP header (X-Conversation-ID)
 * 
 * The backend NEVER sees user email - only the conversation ID!
 */

export interface UserContext {
  email: string;
  beta: boolean;
  prefix: string; // Result of 'conversationid-prefix' flag
  conversationId: string; // Generated ID with prefix: "{prefix}_conv_{uniqueId}"
}

@Injectable({
  providedIn: 'root',
})
export class LaunchDarklyService {
  private ldClient: LDClient.LDClient | null = null;
  private userContext$ = new BehaviorSubject<UserContext | null>(null);

  getUserContext(): Observable<UserContext | null> {
    return this.userContext$.asObservable();
  }

  getCurrentUserContext(): UserContext | null {
    return this.userContext$.value;
  }

  /**
   * 🎯 DEMO STEP 1: User Login
   * Initializes LaunchDarkly and evaluates the conversation-prefix flag
   */
  async initialize(email: string, beta: boolean): Promise<void> {
    if (!environment.ldClientId) {
      console.warn('⚠️ LaunchDarkly client ID not configured. Using default (no prefix).');
      const prefix = '';
      const conversationId = this.generateConversationId(prefix);
      this.userContext$.next({ email, beta, prefix, conversationId });
      return;
    }

    try {
      // Create user context for LaunchDarkly (client-side flag evaluation)
      const user: LDClient.LDContext = {
        kind: 'user',
        key: email,
        email,
        beta,
      };

      console.log('🚀 Initializing LaunchDarkly with user context:', { email, beta });
      this.ldClient = LDClient.initialize(environment.ldClientId, user);
      await this.ldClient.waitForInitialization();

      // 🎯 FLAG #1: Evaluate 'conversationid-prefix'
      // This determines the tier based on LaunchDarkly targeting rules
      const prefix = this.ldClient.variation('conversationid-prefix', '');
      const conversationId = this.generateConversationId(prefix);

      console.log('✅ Flag evaluated: conversationid-prefix =', prefix || '(empty)');
      console.log('📋 Generated conversation ID:', conversationId);

      this.userContext$.next({ email, beta, prefix, conversationId });
    } catch (error) {
      console.error('❌ Error initializing LaunchDarkly:', error);
      const prefix = '';
      const conversationId = this.generateConversationId(prefix);
      this.userContext$.next({ email, beta, prefix, conversationId });
    }
  }

  async switchToAnonymous(): Promise<void> {
    // Generate anonymous conversationID (no prefix)
    const conversationId = this.generateConversationId('');
    this.userContext$.next({ email: '', beta: false, prefix: '', conversationId });

    if (this.ldClient) {
      await this.ldClient.close();
      this.ldClient = null;
    }
  }

  /**
   * Returns the current conversation prefix from the flag evaluation
   */
  getConversationPrefix(): string {
    if (!this.ldClient) {
      return this.userContext$.value?.prefix ?? '';
    }
    return this.ldClient.variation('conversationid-prefix', '') as string;
  }

  /**
   * Optional: Client-side voice flag (not used in this demo)
   * This demo uses SERVER-SIDE flag evaluation instead (see backend)
   */
  isVoiceEnabled(): boolean {
    if (!this.ldClient) {
      return false;
    }
    return this.ldClient.variation('voice-chat-enabled-client', false) as boolean;
  }

  /**
   * 🎯 DEMO: Conversation ID Generation
   * Creates a unique conversation ID with the tier prefix
   * 
   * Examples:
   * - prefix="premium" → "premium_conv_a1b2c3d4"
   * - prefix="test"    → "test_conv_x9y8z7w6"
   * - prefix=""        → "conv_m5n4o3p2" (anonymous)
   */
  private generateConversationId(prefix: string): string {
    const uniqueId = nanoid(8);
    if (prefix) {
      return `${prefix}_conv_${uniqueId}`;
    }
    return `conv_${uniqueId}`;
  }
}
